delete.go

```diff
@@ -66,22 +65,25 @@ status of "ubuntu01" as "stopped" the following will delete resources held for
 			}
 			return err
 		}
+		// When --force is given, we kill all container processes and
+		// then destroy the container. This is done even for a stopped
+		// container, because (in case it does not have its own PID
+		// namespace) there may be some leftover processes in the
+		// container's cgroup.
+		if force {
+			return killContainer(container)
+		}
 		s, err := container.Status()
 		if err != nil {
 			return err
 		}
 		switch s {
 		case libcontainer.Stopped:
-			destroy(container)
+			return container.Destroy()
```

Comment:
I think maybe we can use `killContainer(container)` here, and remove `signalAllProcesses` in `Destory` function.
If yes, all the delete logic will be simplified like this:
```go
if force || s==libcontainer.Stopped || s==libcontainer.Created {
    return killContainer(container)
}
return fmt.Errorf("cannot delete container %s that is not stopped: %s", id, s)
```

Reply:
Yes, I wanted to do it (this will simplify the code), but decided not to; let me explain.

Here is the big picture (took me a while to get).

In this repository we maintain two things:
1. `runc`, as a program.
2. `github.com/opencontainers/runc/libcontainer`, as a set of Go packages.

Out of these packages, the `github.com/opencontainers/runc/libcontainer` itself can be used to create/start/sto/remove containers, pretty much like runc binary, but without the need to exec runc.

Apparently, there are existing users. I am not entirely sure who are they exactly, but we've received some libcontainer related bug reports in the past. At the very least, we have some tests in libcontainer/integration -- this is the only user we're surely aware of.

So, when we do something to fix runc, we should also try to fix it in libcontainer. A libcontainer user can create a container without own private pidns, and can try to delete it when its init is gone but some other processes remain. That's why container.Destroy (which is a part of public libcontainer API) have to (unfortunately) do the killing.